fix: replace scenes in build index with full scene path hash value

NoelStephensUnity opened the pull request:
Migrate from Scenes in Build Index to hash value generated from full scene path.

It was determined that the NetworkSceneManager's implementation/usage of the Scenes in Build indices would create additional user pain in having to include all scenes in the build for both the client and on a server (specifically dedicated server). To avoid this issue, it was suggested/recommended using hash values of the full scene path in place of the scenes in build index values.  This would allow a dedicated server to not be required to have the exact same scenes in build settings and would provide a more robust solution to communicating "which scene" is to be loaded or unloaded.

[MTT-1389](https://jira.unity3d.com/browse/MTT-1389)
[MTT-1286](https://jira.unity3d.com/browse/MTT-1286) 


## Testing and Documentation
* Existing NetworkSceneManager unit tests are still working ( and passing )
* Validation of unique build settings has to be done manually 
* No documentation changes or additions were necessary.

NoelStephensUnity committed:
feat

Migrate from Scenes in Build Index to hash value generated from full scene path.

NoelStephensUnity committed:
style

added comment and removed LF.

NoelStephensUnity committed:
refactor

Removing NetworkSceneManager.GetBuildIndexFromSceneName as it just translated to SceneUtility.GetBuildIndexByScenePath.
Replaced references in NetworkSceneManagerTests with SceneUtility.GetBuildIndexByScenePath.

jeffreyrainy commented on com.unity.netcode.gameobjects/Runtime/SceneManagement/NetworkSceneManager.cs:
```diff
@@ -266,14 +269,72 @@ internal string GetSceneNameFromPath(string scenePath)
+                var scenePath = SceneUtility.GetScenePathByBuildIndex(i);
+                var hash = XXHash.Hash32(scenePath);
+                var buildIndex = SceneUtility.GetBuildIndexByScenePath(scenePath);
+                HashToBuildIndex.Add(hash, buildIndex);
```
There's coupling between the two lookup structures. They should always be consistent with each other. To guarantee that this holds, one could have a `BidirectionalDictionary`. I'm too sure it's a good idea, its extra complexity, but it would guarantee consistency between `HashToBuildIndex` and `BuildIndexToHash`.

Alternatively, we could add for non-ship builds a check done regularly that verify the two mapping are consistent.

NoelStephensUnity replied:
Well, I understand this desire, but there is only one time when this is generated...and it is done all inside that one method.  For V1.0.0. once generated, the intention is that it will not change for the duration of that session.  There is no official BidirectionalDictionary, but it is basically the same thing with added checks if something changes.  These tables, for V1.0,0, should not change.

jeffreyrainy commented on com.unity.netcode.gameobjects/Runtime/SceneManagement/NetworkSceneManager.cs:
```diff
@@ -607,15 +630,15 @@ private SceneEventProgress ValidateSceneEvent(string sceneName, bool isUnloading
 
             // Return invalid scene name status if the scene name is invalid
-            if (!IsSceneNameValid(sceneName))
+            if (SceneUtility.GetBuildIndexByScenePath(sceneName) == -1)
```
Can we get a named constant instead of `-1` ? `InvalidIndex` or `NotFound` or something ?

NoelStephensUnity replied:
Yes! We can... good call out.

NoelStephensUnity committed:
refactor

Making the value to check for an invalid scene name or path (-1) a constant.

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/SceneManagement/NetworkSceneManager.cs:
```diff
@@ -87,6 +87,7 @@ public class SceneEvent
     public class NetworkSceneManager : IDisposable
     {
         private const NetworkDelivery k_DeliveryType = NetworkDelivery.ReliableFragmentedSequenced;
+        internal const int InvalidSceneNameOrPath = -1;
```
nit:
```suggestion
        internal const uint InvalidSceneNameOrPathHash = 0;
```
(updated according to my comment below)

NoelStephensUnity replied:
This is for SceneUtility.GetBuildIndexByScenePath that will return -1 if the scene name or scene path provided is invalid.

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/SceneManagement/NetworkSceneManager.cs:
```diff
@@ -266,14 +270,72 @@ internal string GetSceneNameFromPath(string scenePath)
+        {
+            // In the event there is no scene associated with the scene event
+            // (this can happen) then just return "No Scene";
+            if (sceneHash == 0)
```
what about `InvalidSceneNameOrPath` then? I'm confused. Is `-1` the invalid value or `0`? also hashes are `unsigned`, so better use `uint` for `InvalidSceneNameOrPath` too?

NoelStephensUnity replied:
I updated the comments for better clarity:
            // This can happen during unit tests when clients first connect and the only scene loaded is the 
            // unit test scene (which is ignored by default) that results in a scene event that has no associated
            // scene.  Under this specific special case, we just return "No Scene".

NoelStephensUnity committed:
style

updated comments within the internal string SceneNameFromHash(uint sceneHash) method to better explain why we are checking for a sceneHash value of zero.

mattwalsh-unity committed:
Merge branch 'develop' into fix/hashed-scene-path-replaces-build-index

NoelStephensUnity committed:
Merge branch 'develop' into fix/hashed-scene-path-replaces-build-index

NoelStephensUnity committed:
Merge branch 'develop' into fix/hashed-scene-path-replaces-build-index

NoelStephensUnity committed:
fix

Fixing merge issue.

NoelStephensUnity committed:
Merge branch 'develop' into fix/hashed-scene-path-replaces-build-index

NoelStephensUnity committed:
Merge branch 'develop' into fix/hashed-scene-path-replaces-build-index

andrews-unity committed:
Merge branch 'develop' into fix/hashed-scene-path-replaces-build-index

andrews-unity committed:
Merge branch 'develop' into fix/hashed-scene-path-replaces-build-index

andrews-unity committed:
Merge branch 'develop' into fix/hashed-scene-path-replaces-build-index

NoelStephensUnity committed:
Merge branch 'develop' into fix/hashed-scene-path-replaces-build-index
